File: packages/agent/src/capabilities/tools/fs/edit.ts
/**
 * @fileoverview Edit tool for file editing
 *
 * Performs search and replace operations on files.
 * Returns unified diff output showing exactly what changed.
 */

import * as fs from 'fs/promises';
import type { TronTool, TronToolResult, ToolExecutionOptions } from '@core/types/index.js';
import { createLogger, categorizeError } from '@infrastructure/logging/index.js';
import {
  resolvePath,
  validateRequiredString,
  formatFsError,
} from '../utils.js';

/**
 * Generate a unified diff between old and new strings.
 * Shows context around the change with +/- prefixes.
 */
function generateUnifiedDiff(
  oldStr: string,
  newStr: string,
  contextLines: number = 3
): string {
  const oldLines = oldStr.split('\n');
  const newLines = newStr.split('\n');

  const diffLines: string[] = [];

  // Find the first differing line
  let firstDiff = 0;
  while (
    firstDiff < oldLines.length &&
    firstDiff < newLines.length &&
    oldLines[firstDiff] === newLines[firstDiff]
  ) {
    firstDiff++;
  }

  // Find the last differing line (from the end)
  let oldEnd = oldLines.length - 1;
  let newEnd = newLines.length - 1;
  while (
    oldEnd > firstDiff &&
    newEnd > firstDiff &&
    oldLines[oldEnd] === newLines[newEnd]
  ) {
    oldEnd--;
    newEnd--;
  }

  // Calculate context boundaries
  const contextStart = Math.max(0, firstDiff - contextLines);
  const oldContextEnd = Math.min(oldLines.length - 1, oldEnd + contextLines);

  // Add hunk header
  const oldStart = contextStart + 1;
  const oldCount = oldEnd - contextStart + 1 + Math.min(contextLines, oldLines.length - oldEnd - 1);
  const newStart = contextStart + 1;
  const newCount = newEnd - contextStart + 1 + Math.min(contextLines, newLines.length - newEnd - 1);
  diffLines.push(`@@ -${oldStart},${oldCount} +${newStart},${newCount} @@`);

  // Add context before
  for (let i = contextStart; i < firstDiff; i++) {
    diffLines.push(` ${oldLines[i]}`);
  }

  // Add removed lines
  for (let i = firstDiff; i <= oldEnd; i++) {
    diffLines.push(`-${oldLines[i]}`);
  }

  // Add added lines
  for (let i = firstDiff; i <= newEnd; i++) {
    diffLines.push(`+${newLines[i]}`);
  }

  // Add context after
  const afterStart = oldEnd + 1;
  for (let i = afterStart; i <= oldContextEnd; i++) {
    diffLines.push(` ${oldLines[i]}`);
  }

  return diffLines.join('\n');
}

const logger = createLogger('tool:edit');

export interface EditToolConfig {
  workingDirectory: string;
}

export class EditTool implements TronTool {
  readonly name = 'Edit';
  readonly executionContract = 'options' as const;
  readonly description = 'Edit a file by replacing old_string with new_string. Requires exact match.';
  readonly parameters = {
    type: 'object' as const,
    properties: {
      file_path: {
        type: 'string' as const,
        description: 'The absolute or relative path to the file to edit',
      },
      old_string: {
        type: 'string' as const,
        description: 'The exact string to search for and replace',
      },
      new_string: {
        type: 'string' as const,
        description: 'The string to replace old_string with',
      },
      replace_all: {
        type: 'boolean' as const,
        description: 'Replace all occurrences (default: false)',
        default: false,
      },
    },
    required: ['file_path', 'old_string', 'new_string'] as string[],
  };

  private config: EditToolConfig;

  constructor(config: EditToolConfig) {
    this.config = config;
  }

  async execute(args: Record<string, unknown>, _options?: ToolExecutionOptions): Promise<TronToolResult> {
    // Validate required parameters
    const pathValidation = validateRequiredString(args, 'file_path', 'the path to the file to edit');
    if (!pathValidation.valid) return pathValidation.error!;

    if (args.old_string === undefined || args.old_string === null) {
      return {
        content: 'Missing required parameter: old_string. The tool call may have been truncated.',
        isError: true,
      };
    }
    if (args.new_string === undefined || args.new_string === null) {
      return {
        content: 'Missing required parameter: new_string. The tool call may have been truncated.',
        isError: true,
      };
    }

    const filePath = resolvePath(args.file_path as string, this.config.workingDirectory);
    const oldString = args.old_string as string;
    const newString = args.new_string as string;
    const replaceAll = (args.replace_all as boolean) ?? false;

    const startTime = Date.now();
    logger.debug('Editing file', { filePath, replaceAll });

    // Validate inputs
    if (oldString === newString) {
      return {
        content: 'Error: old_string and new_string are the same. No changes needed.',
        isError: true,
        details: { filePath },
      };
    }

    try {
      const content = await fs.readFile(filePath, 'utf-8');

      // Count occurrences
      const occurrences = this.countOccurrences(content, oldString);

      if (occurrences === 0) {
        return {
          content: `Error: old_string not found in file. The exact string "${this.truncate(oldString, 50)}" does not exist in ${filePath}`,
          isError: true,
          details: { filePath, occurrences: 0 },
        };
      }

      // If multiple occurrences and not replace_all, error
      if (occurrences > 1 && !replaceAll) {
        return {
          content: `Error: old_string appears multiple times (${occurrences} occurrences). Use replace_all: true to replace all occurrences, or provide more context to make the match unique.`,
          isError: true,
          details: { filePath, occurrences },
        };
      }

      // Perform replacement
      let newContent: string;
      let replacements: number;

      if (replaceAll) {
        newContent = content.split(oldString).join(newString);
        replacements = occurrences;
      } else {
        newContent = content.replace(oldString, newString);
        replacements = 1;
      }

      // Write the file
      await fs.writeFile(filePath, newContent, 'utf-8');

      const duration = Date.now() - startTime;
      logger.info('File edit completed', {
        filePath,
        replacements,
        oldStringLength: oldString.length,
        newStringLength: newString.length,
        duration,
      });

      // Generate unified diff for display
      const diff = generateUnifiedDiff(oldString, newString, 2);

      // Build result content with diff
      const resultContent = [
        `Successfully replaced ${replacements} occurrence${replacements > 1 ? 's' : ''} in ${filePath}`,
        '',
        diff,
      ].join('\n');

      return {
        content: resultContent,
        isError: false,
        details: {
          filePath,
          replacements,
          oldStringPreview: this.truncate(oldString, 50),
          newStringPreview: this.truncate(newString, 50),
          diff,
        },
      };
    } catch (error) {
      const duration = Date.now() - startTime;
      const structuredError = categorizeError(error, { filePath, operation: 'edit' });
      logger.error('File edit failed', {
        filePath,
        error: structuredError.message,
        code: structuredError.code,
        category: structuredError.category,
        duration,
      });
      return formatFsError(error, filePath, 'editing');
    }
  }

  private countOccurrences(str: string, search: string): number {
    let count = 0;
    let pos = 0;
    while ((pos = str.indexOf(search, pos)) !== -1) {
      count++;
      pos += search.length;
    }
    return count;
  }

  private truncate(str: string, maxLen: number): string {
    if (str.length <= maxLen) return str;
    return str.substring(0, maxLen) + '...';
  }
}

File: packages/agent-rs/crates/tron-tools/src/fs/edit.rs
//! `Edit` tool — exact string replacement with diff output.
//!
//! Finds an exact substring in a file and replaces it. Returns a unified diff
//! showing the change. Supports `replace_all` for multiple occurrences.

use std::sync::Arc;

use async_trait::async_trait;
use serde_json::{json, Value};
use tron_core::tools::{
    Tool, ToolCategory, ToolParameterSchema, ToolResultBody, TronToolResult, error_result,
};

use crate::errors::ToolError;
use crate::traits::{FileSystemOps, ToolContext, TronTool};
use crate::utils::diff::generate_unified_diff;
use crate::utils::fs_errors::format_fs_error;
use crate::utils::path::resolve_path;
use crate::utils::validation::validate_required_string;

/// The `Edit` tool performs exact string replacement in files.
pub struct EditTool {
    fs: Arc<dyn FileSystemOps>,
}

impl EditTool {
    /// Create a new `Edit` tool with the given filesystem.
    pub fn new(fs: Arc<dyn FileSystemOps>) -> Self {
        Self { fs }
    }
}

fn truncate_preview(s: &str, max_len: usize) -> String {
    // max_len is the body limit; total output may be up to max_len + suffix.
    let total = max_len.saturating_add(3);
    tron_core::text::truncate_with_suffix(s, total, "...")
}

#[async_trait]
impl TronTool for EditTool {
    fn name(&self) -> &str {
        "Edit"
    }

    fn category(&self) -> ToolCategory {
        ToolCategory::Filesystem
    }

    fn definition(&self) -> Tool {
        Tool {
            name: "Edit".into(),
            description: "Edit a file by replacing old_string with new_string. Requires exact match.".into(),
            parameters: ToolParameterSchema {
                schema_type: "object".into(),
                properties: Some({
                    let mut m = serde_json::Map::new();
                    let _ = m.insert(
                        "file_path".into(),
                        json!({"type": "string", "description": "The path to the file to edit"}),
                    );
                    let _ = m.insert(
                        "old_string".into(),
                        json!({"type": "string", "description": "The exact string to find and replace"}),
                    );
                    let _ = m.insert(
                        "new_string".into(),
                        json!({"type": "string", "description": "The replacement string"}),
                    );
                    let _ = m.insert(
                        "replace_all".into(),
                        json!({"type": "boolean", "description": "Replace all occurrences (default: false)"}),
                    );
                    m
                }),
                required: Some(vec![
                    "file_path".into(),
                    "old_string".into(),
                    "new_string".into(),
                ]),
                description: None,
                extra: serde_json::Map::new(),
            },
        }
    }

    async fn execute(
        &self,
        params: Value,
        ctx: &ToolContext,
    ) -> Result<TronToolResult, ToolError> {
        let file_path = match validate_required_string(&params, "file_path", "path to the file") {
            Ok(p) => p,
            Err(e) => return Ok(e),
        };

        let old_string = match params.get("old_string") {
            Some(Value::String(s)) => s.clone(),
            _ => {
                return Ok(error_result(
                    "Missing required parameter: old_string (the string to find)",
                ));
            }
        };

        let new_string = match params.get("new_string") {
            Some(Value::String(s)) => s.clone(),
            _ => {
                return Ok(error_result(
                    "Missing required parameter: new_string (the replacement string)",
                ));
            }
        };

        if old_string.is_empty() {
            return Ok(error_result("old_string must not be empty"));
        }

        if old_string == new_string {
            return Ok(error_result(
                "old_string and new_string are identical — no changes would be made",
            ));
        }

        let replace_all = params
            .get("replace_all")
            .and_then(Value::as_bool)
            .unwrap_or(false);

        let resolved = resolve_path(&file_path, &ctx.working_directory);

        // Read the file
        let bytes = match self.fs.read_file(&resolved).await {
            Ok(b) => b,
            Err(e) => return Ok(format_fs_error(&e, &resolved.to_string_lossy(), "reading")),
        };

        let content = String::from_utf8_lossy(&bytes).into_owned();

        // Count occurrences
        let count = content.matches(&old_string).count();
        if count == 0 {
            let preview = truncate_preview(&old_string, 50);
            return Ok(error_result(format!(
                "old_string not found in file: \"{preview}\""
            )));
        }

        if count > 1 && !replace_all {
            return Ok(error_result(format!(
                "Found {count} occurrences of old_string. Use replace_all: true to replace all, or make old_string more specific."
            )));
        }

        // Perform replacement
        let new_content = if replace_all {
            content.replace(&old_string, &new_string)
        } else {
            content.replacen(&old_string, &new_string, 1)
        };

        let replacements = if replace_all { count } else { 1 };

        // Generate diff
        let diff = generate_unified_diff(&content, &new_content, 3);

        // Write the modified file
        if let Err(e) = self.fs.write_file(&resolved, new_content.as_bytes()).await {
            return Ok(format_fs_error(&e, &resolved.to_string_lossy(), "writing"));
        }

        let details = json!({
            "filePath": resolved.to_string_lossy(),
            "replacements": replacements,
            "oldStringPreview": truncate_preview(&old_string, 50),
            "newStringPreview": truncate_preview(&new_string, 50),
            "diff": diff,
        });

        Ok(TronToolResult {
            content: ToolResultBody::Blocks(vec![
                tron_core::content::ToolResultContent::text(diff),
            ]),
            details: Some(details),
            is_error: None,
            stop_turn: None,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashMap;
    use std::io;
    use std::path::{Path, PathBuf};
    use std::sync::Mutex;

    struct MockFs {
        files: Mutex<HashMap<PathBuf, Vec<u8>>>,
    }

    impl MockFs {
        fn new() -> Self {
            Self { files: Mutex::new(HashMap::new()) }
        }
        fn with_file(self, path: impl Into<PathBuf>, content: impl AsRef<[u8]>) -> Self {
            let _ = self.files.lock().unwrap().insert(path.into(), content.as_ref().to_vec());
            self
        }
        fn read_content(&self, path: &Path) -> Option<String> {
            self.files.lock().unwrap().get(path).map(|b| String::from_utf8_lossy(b).into_owned())
        }
    }

    #[async_trait]
    impl FileSystemOps for MockFs {
        async fn read_file(&self, path: &Path) -> Result<Vec<u8>, io::Error> {
            self.files.lock().unwrap().get(path).cloned()
                .ok_or_else(|| io::Error::new(io::ErrorKind::NotFound, "not found"))
        }
        async fn write_file(&self, path: &Path, content: &[u8]) -> Result<(), io::Error> {
            let _ = self.files.lock().unwrap().insert(path.to_path_buf(), content.to_vec());
            Ok(())
        }
        async fn metadata(&self, _: &Path) -> Result<std::fs::Metadata, io::Error> {
            Err(io::Error::new(io::ErrorKind::Other, "mock"))
        }
        async fn create_dir_all(&self, _: &Path) -> Result<(), io::Error> { Ok(()) }
        fn exists(&self, path: &Path) -> bool {
            self.files.lock().unwrap().contains_key(path)
        }
    }

    fn make_ctx() -> ToolContext {
        ToolContext {
            tool_call_id: "call-1".into(),
            session_id: "sess-1".into(),
            working_directory: "/tmp".into(),
            cancellation: tokio_util::sync::CancellationToken::new(),
            subagent_depth: 0,
            subagent_max_depth: 0,
        }
    }

    fn extract_text(result: &TronToolResult) -> String {
        match &result.content {
            ToolResultBody::Text(t) => t.clone(),
            ToolResultBody::Blocks(blocks) => blocks.iter().filter_map(|b| match b {
                tron_core::content::ToolResultContent::Text { text } => Some(text.as_str()),
                _ => None,
            }).collect::<Vec<_>>().join(""),
        }
    }

    #[tokio::test]
    async fn exact_match_replace() {
        let fs = Arc::new(MockFs::new().with_file("/tmp/f.txt", "hello world"));
        let tool = EditTool::new(fs.clone());
        let result = tool.execute(
            json!({"file_path": "/tmp/f.txt", "old_string": "hello", "new_string": "goodbye"}),
            &make_ctx(),
        ).await.unwrap();
        assert!(result.is_error.is_none());
        assert_eq!(fs.read_content(Path::new("/tmp/f.txt")).unwrap(), "goodbye world");
        let text = extract_text(&result);
        assert!(text.contains("-hello"));
        assert!(text.contains("+goodbye"));
    }

    #[tokio::test]
    async fn old_string_not_found() {
        let fs = Arc::new(MockFs::new().with_file("/tmp/f.txt", "hello"));
        let tool = EditTool::new(fs);
        let result = tool.execute(
            json!({"file_path": "/tmp/f.txt", "old_string": "xyz", "new_string": "abc"}),
            &make_ctx(),
        ).await.unwrap();
        assert_eq!(result.is_error, Some(true));
        let text = extract_text(&result);
        assert!(text.contains("not found"));
    }

    #[tokio::test]
    async fn multiple_without_replace_all() {
        let fs = Arc::new(MockFs::new().with_file("/tmp/f.txt", "aaa bbb aaa"));
        let tool = EditTool::new(fs);
        let result = tool.execute(
            json!({"file_path": "/tmp/f.txt", "old_string": "aaa", "new_string": "xxx"}),
            &make_ctx(),
        ).await.unwrap();
        assert_eq!(result.is_error, Some(true));
        let text = extract_text(&result);
        assert!(text.contains("2 occurrences"));
    }

    #[tokio::test]
    async fn replace_all_multiple() {
        let fs = Arc::new(MockFs::new().with_file("/tmp/f.txt", "aaa bbb aaa"));
        let tool = EditTool::new(fs.clone());
        let result = tool.execute(
            json!({"file_path": "/tmp/f.txt", "old_string": "aaa", "new_string": "xxx", "replace_all": true}),
            &make_ctx(),
        ).await.unwrap();
        assert!(result.is_error.is_none());
        assert_eq!(fs.read_content(Path::new("/tmp/f.txt")).unwrap(), "xxx bbb xxx");
        let details = result.details.unwrap();
        assert_eq!(details["replacements"], 2);
    }

    #[tokio::test]
    async fn identical_strings_error() {
        let fs = Arc::new(MockFs::new().with_file("/tmp/f.txt", "hello"));
        let tool = EditTool::new(fs);
        let result = tool.execute(
            json!({"file_path": "/tmp/f.txt", "old_string": "hello", "new_string": "hello"}),
            &make_ctx(),
        ).await.unwrap();
        assert_eq!(result.is_error, Some(true));
        let text = extract_text(&result);
        assert!(text.contains("identical"));
    }

    #[tokio::test]
    async fn empty_old_string() {
        let fs = Arc::new(MockFs::new().with_file("/tmp/f.txt", "hello"));
        let tool = EditTool::new(fs);
        let result = tool.execute(
            json!({"file_path": "/tmp/f.txt", "old_string": "", "new_string": "x"}),
            &make_ctx(),
        ).await.unwrap();
        assert_eq!(result.is_error, Some(true));
    }

    #[tokio::test]
    async fn multiline_replacement() {
        let content = "line1\nline2\nline3\n";
        let fs = Arc::new(MockFs::new().with_file("/tmp/f.txt", content));
        let tool = EditTool::new(fs.clone());
        let result = tool.execute(
            json!({"file_path": "/tmp/f.txt", "old_string": "line2\nline3", "new_string": "new2\nnew3"}),
            &make_ctx(),
        ).await.unwrap();
        assert!(result.is_error.is_none());
        assert_eq!(fs.read_content(Path::new("/tmp/f.txt")).unwrap(), "line1\nnew2\nnew3\n");
    }

    #[tokio::test]
    async fn unicode_replacement() {
        let fs = Arc::new(MockFs::new().with_file("/tmp/f.txt", "caf\u{00E9} \u{1F600}"));
        let tool = EditTool::new(fs.clone());
        let result = tool.execute(
            json!({"file_path": "/tmp/f.txt", "old_string": "caf\u{00E9}", "new_string": "coffee"}),
            &make_ctx(),
        ).await.unwrap();
        assert!(result.is_error.is_none());
        assert_eq!(fs.read_content(Path::new("/tmp/f.txt")).unwrap(), "coffee \u{1F600}");
    }

    #[tokio::test]
    async fn diff_output_format() {
        let fs = Arc::new(MockFs::new().with_file("/tmp/f.txt", "a\nb\nc\nd\ne"));
        let tool = EditTool::new(fs);
        let result = tool.execute(
            json!({"file_path": "/tmp/f.txt", "old_string": "c", "new_string": "C"}),
            &make_ctx(),
        ).await.unwrap();
        let text = extract_text(&result);
        assert!(text.contains("@@"));
        assert!(text.contains("-c"));
        assert!(text.contains("+C"));
    }

    #[tokio::test]
    async fn file_not_found() {
        let fs = Arc::new(MockFs::new());
        let tool = EditTool::new(fs);
        let result = tool.execute(
            json!({"file_path": "/tmp/missing.txt", "old_string": "a", "new_string": "b"}),
            &make_ctx(),
        ).await.unwrap();
        assert_eq!(result.is_error, Some(true));
    }

    #[tokio::test]
    async fn missing_file_path() {
        let fs = Arc::new(MockFs::new());
        let tool = EditTool::new(fs);
        let result = tool.execute(
            json!({"old_string": "a", "new_string": "b"}),
            &make_ctx(),
        ).await.unwrap();
        assert_eq!(result.is_error, Some(true));
    }

    #[tokio::test]
    async fn missing_old_string() {
        let fs = Arc::new(MockFs::new());
        let tool = EditTool::new(fs);
        let result = tool.execute(
            json!({"file_path": "/tmp/f.txt", "new_string": "b"}),
            &make_ctx(),
        ).await.unwrap();
        assert_eq!(result.is_error, Some(true));
    }

    #[tokio::test]
    async fn missing_new_string() {
        let fs = Arc::new(MockFs::new());
        let tool = EditTool::new(fs);
        let result = tool.execute(
            json!({"file_path": "/tmp/f.txt", "old_string": "a"}),
            &make_ctx(),
        ).await.unwrap();
        assert_eq!(result.is_error, Some(true));
    }

    #[tokio::test]
    async fn details_include_replacements_and_diff() {
        let fs = Arc::new(MockFs::new().with_file("/tmp/f.txt", "hello world"));
        let tool = EditTool::new(fs);
        let result = tool.execute(
            json!({"file_path": "/tmp/f.txt", "old_string": "hello", "new_string": "goodbye"}),
            &make_ctx(),
        ).await.unwrap();
        let details = result.details.unwrap();
        assert_eq!(details["replacements"], 1);
        assert!(details["diff"].as_str().unwrap().contains("@@"));
    }

    #[tokio::test]
    async fn whitespace_differences_detected() {
        let fs = Arc::new(MockFs::new().with_file("/tmp/f.txt", "hello  world"));
        let tool = EditTool::new(fs.clone());
        let result = tool.execute(
            json!({"file_path": "/tmp/f.txt", "old_string": "hello  world", "new_string": "hello world"}),
            &make_ctx(),
        ).await.unwrap();
        assert!(result.is_error.is_none());
        assert_eq!(fs.read_content(Path::new("/tmp/f.txt")).unwrap(), "hello world");
    }

    #[tokio::test]
    async fn tab_chars_preserved() {
        let fs = Arc::new(MockFs::new().with_file("/tmp/f.txt", "\thello\t"));
        let tool = EditTool::new(fs.clone());
        let result = tool.execute(
            json!({"file_path": "/tmp/f.txt", "old_string": "\thello\t", "new_string": "  hello  "}),
            &make_ctx(),
        ).await.unwrap();
        assert!(result.is_error.is_none());
        assert_eq!(fs.read_content(Path::new("/tmp/f.txt")).unwrap(), "  hello  ");
    }

    #[tokio::test]
    async fn replace_in_first_line() {
        let fs = Arc::new(MockFs::new().with_file("/tmp/f.txt", "FIRST\nsecond\nthird"));
        let tool = EditTool::new(fs.clone());
        let result = tool.execute(
            json!({"file_path": "/tmp/f.txt", "old_string": "FIRST", "new_string": "first"}),
            &make_ctx(),
        ).await.unwrap();
        assert!(result.is_error.is_none());
        let content = fs.read_content(Path::new("/tmp/f.txt")).unwrap();
        assert!(content.starts_with("first\n"));
    }

    #[tokio::test]
    async fn replace_in_last_line() {
        let fs = Arc::new(MockFs::new().with_file("/tmp/f.txt", "first\nsecond\nLAST"));
        let tool = EditTool::new(fs.clone());
        let result = tool.execute(
            json!({"file_path": "/tmp/f.txt", "old_string": "LAST", "new_string": "last"}),
            &make_ctx(),
        ).await.unwrap();
        assert!(result.is_error.is_none());
        let content = fs.read_content(Path::new("/tmp/f.txt")).unwrap();
        assert!(content.ends_with("last"));
    }

    #[tokio::test]
    async fn large_file_single_replacement() {
        let lines: Vec<String> = (1..=10_000).map(|i| format!("line {i}")).collect();
        let content = lines.join("\n");
        let fs = Arc::new(MockFs::new().with_file("/tmp/big.txt", content.as_str()));
        let tool = EditTool::new(fs.clone());
        let result = tool.execute(
            json!({"file_path": "/tmp/big.txt", "old_string": "line 5000", "new_string": "REPLACED"}),
            &make_ctx(),
        ).await.unwrap();
        assert!(result.is_error.is_none());
        let new_content = fs.read_content(Path::new("/tmp/big.txt")).unwrap();
        assert!(new_content.contains("REPLACED"));
        assert!(!new_content.contains("line 5000"));
    }
}
